http/src/main/scala/org/http4s/blaze/http/http2/Http2Exception.scala

```diff
@@ -0,0 +1,112 @@
+package org.http4s.blaze.http.http2
+
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets.UTF_8
+import scala.collection.mutable
+import scala.util.control.NoStackTrace
+
+sealed abstract class Http2Exception(msg: String) extends Exception(msg) with NoStackTrace {
+
+  // A 32-bit unsigned Integer
+  def code: Long
+
+  final def name: String = Http2Exception.errorName(code)
+
+  /** Convert this exception to a stream exception
+    *
+    * @note If this is already a stream exception but with a different stream id, the id will be changed
+    */
+  final def toStreamException(streamId: Int): Http2StreamException = this match {
+    case ex: Http2StreamException if ex.stream == streamId => ex
+    case ex => new Http2StreamException(streamId, ex.code, ex.getMessage)
+  }
+
+  /** Convert this exception to a session exception */
+  final def toSessionException(): Http2SessionException = this match {
+    case Http2StreamException(_, code, msg) => Http2SessionException(code, msg)
+    case ex: Http2SessionException => ex
+  }
+
+  /** Was the exception due to refusal by the peer.
+    *
+    * These exceptions are safe to automatically retry even if the HTTP method
+    * is not an idempotent method. See https://tools.ietf.org/html/rfc7540#section-8.1.4
+    * for more details.
+    */
+  final def isRefusedStream: Boolean = code == Http2Exception.REFUSED_STREAM.code
+
+  /** serialize the message as a `ByteBuffer` */
+  final def msgBuffer(): ByteBuffer = ByteBuffer.wrap(msg.getBytes(UTF_8))
+}
+
+final case class Http2StreamException(stream: Int, code: Long, msg: String) extends Http2Exception(msg)
+
+final case class Http2SessionException(code: Long, msg: String) extends Http2Exception(msg)
+
+///////////////////// HTTP/2.0 Errors //////////////////////////////
+object Http2Exception {
+
+  final class ErrorGenerator private[http2](val code: Long, val name: String) {
+    /** Create a Http2Exception with stream id 0 */
+    def goaway(): Http2Exception = Http2SessionException(code, name)
+
+    /** Create a Http2Exception with stream id 0 */
+    def goaway(msg: String): Http2Exception = Http2SessionException(code, name + ": " + msg)
+
+    /** Create a Http2Exception with the requisite stream id */
+    def rst(stream: Int): Http2Exception = rst(stream, name)
+
+    /** Create a Http2Exception with the requisite stream id */
+    def rst(stream: Int, msg: String): Http2Exception = Http2StreamException(stream, code, msg)
+
+    /** Extract the optional stream id and the exception message */
+    def unapply(ex: Http2Exception): Option[(Option[Int], String)] = {
+      if (ex.code == code) {
+        val stream = ex match {
+          case ex: Http2StreamException => Some(ex.stream)
+          case _ => None
+        }
+        Some(stream -> ex.getMessage)
+      }
+      else None
+    }
+
+    def unapply(code: Int): Option[Unit] = {
```

Comment:
Minor, maybe not wanted, but don't forget `unapply` can return `Boolean`.

Reply:
TIL.